### PR TITLE
Work in progress

### PR DESCRIPTION
**Pull request by jenshenneberg**

### CODE REVIEW

**Comment by jenshenneberg** [comment]
TODO:

- [ ] Add readme.md
- [ ] More tests of context mapping

File: test/OpenFeature.Contrib.Providers.Statsig.Test/StatsigProviderTest.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
+using AutoFixture.Xunit2;
+using OpenFeature.Constant;
+using OpenFeature.Model;
+using Statsig.Server;
+using System.Threading.Tasks;
+using Xunit;
+
+
```
Comment (Mackey780) [inline test/OpenFeature.Contrib.Providers.Statsig.Test/StatsigProviderTest.cs:8]: _nit_
```suggestion

```

File: test/OpenFeature.Contrib.Providers.Statsig.Test/StatsigProviderTest.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            new StatsigProvider("secret-", x => x.LocalMode = true);
+    }
+
+    [Fact]
+    public async Task StatsigProvider_Initialized_HasCorrectStatusAsync()
+    {
+        Assert.Equal(ProviderStatus.NotReady, statsigProvider.GetStatus());
+        await statsigProvider.Initialize(null);
+        Assert.Equal(ProviderStatus.Ready, statsigProvider.GetStatus());
+
```
Comment (Mackey780) [inline test/OpenFeature.Contrib.Providers.Statsig.Test/StatsigProviderTest.cs:27]: _nit_
```suggestion
```

**Review by Mackey780 — COMMENTED** [review]
Mostly just nitpicks sorry :)

File: src/OpenFeature.Contrib.Providers.Statsig/EvaluationContextExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        public static StatsigUser AsStatsigUser(this EvaluationContext evaluationContext)
+        {
+            if (evaluationContext == null)
+                return null;
+
+            var user = new StatsigUser();
+            foreach (var item in evaluationContext)
+            {
+                //TODO: Await release containing this https://github.com/open-feature/dotnet-sdk/pull/231 to use TargetingKey instead of UserId
+                if (item.Key == "UserID")
```
[the remark this answers is not in the document]
Comment (jenshenneberg) [reply src/OpenFeature.Contrib.Providers.Statsig/EvaluationContextExtensions.cs]: Changed to use `item.Key.ToUpperInvariant()` so either should work now 😃

[the remark this answers is not in the document]
Comment (Mackey780) [reply src/OpenFeature.Contrib.Providers.Statsig/EvaluationContextExtensions.cs]: Sweet, also used in:
`EvaluationContextExtensionsTests.cs` line 15 & 103
`StatsigProviderTest.cs` line 33

File: src/OpenFeature.Contrib.Providers.Statsig/README.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+| EvaluationContext Key | Statsig User Parameter    |
+|-----------------------|---------------------------|
+| `appVersion`          | `AppVersion`              |
+| `country`             | `Country`                 |
+| `email`               | `Email`                   |
+| `ip`                  | `Ip`                      |
+| `locale`              | `Locale`                  |
+| `userAgent`           | `UserAgent`               |
+| `privateAttributes`   | `PrivateAttributes`       |
```
Comment (benjamincburns) [inline src/OpenFeature.Contrib.Providers.Statsig/README.md:98]: Is this the set of keys that are common between the two user representations? From the statsig docs page you linked, it appears there are quite a few more keys available on the `StatsigUser` object that would be useful for flag evaluation.

If we were indicating org membership in the `StatsigUser` object would that be under `CustomIDs`? If so, perhaps it would make sense to include an extension point for mapping application-specific `EvaluationContext` keys over to the `PrivateAttributes` and `CustomIds` properties.

File: src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        /// <inheritdoc/>
+        public override Task<ResolutionDetails<bool>> ResolveBooleanValue(string flagKey, bool defaultValue, EvaluationContext context = null)
+        {
+            //TODO: defaultvalue = true not yet supported due to https://github.com/statsig-io/dotnet-sdk/issues/33
+            if (defaultValue == true)
+                throw new StatsigProviderException("defaultvalue = true not supported (https://github.com/statsig-io/dotnet-sdk/issues/33)");
+            if (GetStatus() != ProviderStatus.Ready)
+                return Task.FromResult(new ResolutionDetails<bool>(flagKey, defaultValue, ErrorType.ProviderNotReady));
+            var result = StatsigServer.CheckGateSync(context.AsStatsigUser(), flagKey);
+            return Task.FromResult(new ResolutionDetails<bool>(flagKey, result));
```
Comment (benjamincburns) [inline src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs:57]: I could be mistaken, but I think if you make the method as `async` you can drop the `Task.FromResult` boilerplate. It'll probably introduce a warning about an async method with no `await`, however.

File: src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        public override Metadata GetMetadata() => _providerMetadata;
+
+        /// <inheritdoc/>
+        public override Task<ResolutionDetails<bool>> ResolveBooleanValue(string flagKey, bool defaultValue, EvaluationContext context = null)
+        {
+            //TODO: defaultvalue = true not yet supported due to https://github.com/statsig-io/dotnet-sdk/issues/33
+            if (defaultValue == true)
+                throw new StatsigProviderException("defaultvalue = true not supported (https://github.com/statsig-io/dotnet-sdk/issues/33)");
+            if (GetStatus() != ProviderStatus.Ready)
+                return Task.FromResult(new ResolutionDetails<bool>(flagKey, defaultValue, ErrorType.ProviderNotReady));
```
Comment (benjamincburns) [inline src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs:55]: Minor style nitpick - maintenance of one line branches is less error-prone when using blocks/curly braces.

File: src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+            var result = StatsigServer.CheckGateSync(context.AsStatsigUser(), flagKey);
+            return Task.FromResult(new ResolutionDetails<bool>(flagKey, result));
+        }
+
+        /// <inheritdoc/>
+        public override Task<ResolutionDetails<double>> ResolveDoubleValue(string flagKey, double defaultValue, EvaluationContext context = null)
+        {
+            throw new NotImplementedException();
```
Comment (benjamincburns) [inline src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs:63]: Are these not implemented because statsig doesn't support these operations? If so, should it be `InvalidOperationException` with a message that indicates that resolution of the type in question isn't supported by statsig?

Comment (benjamincburns) [reply src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs:63]: The fact that these other types can be resolved by OpenFeature also begs for an implementation of `IConfigurationProvider`. Not that `IConfigurationProvider` supports types other than string, but it's clear that OpenFeature is meant to resolve more complex configuration. Also it's nice to be able to bind to `Options` objects via [the options pattern](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-8.0)

File: src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        {
+            return initialized ? ProviderStatus.Ready : ProviderStatus.NotReady;
+        }
+
+        /// <inheritdoc/>
+        public override async Task Initialize(EvaluationContext context)
+        {
+            if (!initialized)
+            {
+                await semaphore.WaitAsync();
```
Comment (benjamincburns) [inline src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs:95]: Shouldn't the lock be acquired before the `initialized` check? Otherwise if e.g. there are two concurrent calls to `Initialize` the blocked call will just wait until the initialization is complete and then duplicate the initialization work from the first call.

Also it looks like you've configured `SemaphoreSlim` to behave as a mutex here. Perhaps it'd be more idiomatic to replace it with a `lock(this) { ... }`?

File: src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+                {
+                    semaphore.Release();
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public override Task Shutdown()
+        {
+            if (initialized)
```
Comment (benjamincburns) [inline src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs:116]: Should this also acquire the lock, in case we're in the process of initializing while `Shutdown` is called?

File: src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    /// <example>
+    ///     var provider = new StatsigProvider("my-sdk-key"), new StatsigProviderOptions(){LocalMode = false});
+    ///
+    ///     OpenFeature.Api.Instance.SetProvider(provider);
+    ///
+    ///     var client = OpenFeature.Api.Instance.GetClient();
+    /// </example>
+    public sealed class StatsigProvider : FeatureProvider
+    {
+        bool initialized = false;
```
Comment (benjamincburns) [inline src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs:24]: Given the concurrency/locking checks in this class, I think this field should be marked as `volatile`.

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/volatile

File: src/OpenFeature.Contrib.Providers.Statsig/StatsigProviderException.cs [hunk 11]
```diff
@@ -0,0 +1,18 @@
+using System;
+
+namespace OpenFeature.Contrib.Providers.Statsig
+{
+    /// <summary>
+    ///   Generic Statsig provider exception
+    /// </summary>
+    public class StatsigProviderException : Exception
```
Comment (benjamincburns) [inline src/OpenFeature.Contrib.Providers.Statsig/StatsigProviderException.cs:8]: As the code grows, it will be tempting for future maintainers to use this as a generic "something broke, you figure it out" error type when it could be more useful to consumer code to throw more specific errors. That tends to make it cumbersome to branch to the correct error handling logic in consumer code.

I was also hopeful that there would be some `ProviderException` type provided by OpenFeature, but from looking at the other provider implementations there doesn't seem to be one. Needing to have provider-specific error handling code in the consumer codebase means that OpenFeature is likely to be a rather leaky abstraction.

Given that we're only throwing this in a place where we haven't implemented something due to it being unsupported, is `NotImplementedException` or `InvalidOperationException` perhaps more idiomatic?

File: src/OpenFeature.Contrib.Providers.Statsig/README.md (hunk 4 above)
Comment (benjamincburns) [reply src/OpenFeature.Contrib.Providers.Statsig/README.md:98]: Looks like you're mapping `userId` but it's not included in this list.

File: test/OpenFeature.Contrib.Providers.Statsig.Test/EvaluationContextExtensionsTests.cs [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        public void AsStatsigUser_ShouldHandleEmptyUserID()
+        {
+            // Arrange
+            var evaluationContext = EvaluationContext.Builder().Set("UserID", "").Build();
+
+            // Act
+            var statsigUser = evaluationContext.AsStatsigUser();
+
+            // Assert
+            Assert.NotNull(statsigUser);
```
Comment (benjamincburns) [inline test/OpenFeature.Contrib.Providers.Statsig.Test/EvaluationContextExtensionsTests.cs:109]: Should this assert that statsigUser is `""`?

File: test/OpenFeature.Contrib.Providers.Statsig.Test/EvaluationContextExtensionsTests.cs [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+        public void AsStatsigUser_ShouldHandleNullEvaluationContext()
+        {
+            // Arrange
+            EvaluationContext evaluationContext = null;
+
+            // Act
+            var statsigUser = evaluationContext.AsStatsigUser();
+
+            // Assert
+            Assert.Null(statsigUser);
```
Comment (benjamincburns) [inline test/OpenFeature.Contrib.Providers.Statsig.Test/EvaluationContextExtensionsTests.cs:96]: Curious as to why this is desirable. If the eval context is null, why isn't it more helpful to throw rather than to return a null value for `statsigUser`?

**Review by benjamincburns — CHANGES_REQUESTED** [review]
I had a bunch of comments/questions, but marked as "request changes" specifically for the example code fix in the readme and the locking logic in `StatsigProvider`.

File: src/OpenFeature.Contrib.Providers.Statsig/README.md [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+            var statsigProvider = new StatsigProvider("#YOUR-SDK-KEY#");
+
+            // Set the statsigProvider as the provider for the OpenFeature SDK
+            OpenFeature.Api.Instance.SetProvider(statsigProvider);
+
+            var client = OpenFeature.Api.Instance.GetClient();
+
+            var val = client.GetBooleanValue("isMyAwesomeFeatureEnabled", false);
+
+            if(isMyAwesomeFeatureEnabled)
```
[the remark this answers is not in the document]
Comment (benjamincburns) [reply src/OpenFeature.Contrib.Providers.Statsig/README.md:61]: Also I tend to find examples easier to understand when they avoid use of `var` unless the type is obvious from use.

File: src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs (hunk 9 above)
Comment (benjamincburns) [reply src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs:116]: Given that we have locking here, perhaps a concurrency test is in order?

File: src/OpenFeature.Contrib.Providers.Statsig/EvaluationContextExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        //These keys match the keys of the statsiguser object as descibed here
+        //https://docs.statsig.com/client/concepts/user
+        internal const string CONTEXT_APP_VERSION = "appVersion";
+        internal const string CONTEXT_COUNTRY = "country";
+        internal const string CONTEXT_EMAIL = "email";
+        internal const string CONTEXT_IP = "ip";
+        internal const string CONTEXT_LOCALE = "locale";
+        internal const string CONTEXT_USER_AGENT = "userAgent";
+        internal const string CONTEXT_PRIVATE_ATTRIBUTES = "privateAttributes";
```
[the remark this answers is not in the document]
Comment (jenshenneberg) [reply src/OpenFeature.Contrib.Providers.Statsig/EvaluationContextExtensions.cs:16]: As you mention below, I'd still need to loop through the list to add custom properties using `AddCustomProperty`. I'll have a look to see if the mapping code can be done more elegantly.

File: src/OpenFeature.Contrib.Providers.Statsig/README.md [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+
+```shell
+// Install OpenFeature.Contrib.Providers.Statsig as a Cake Addin
+#addin nuget:?package=OpenFeature.Contrib.Providers.Statsig
+
+// Install OpenFeature.Contrib.Providers.Statsig as a Cake Tool
+#tool nuget:?package=OpenFeature.Contrib.Providers.Statsig
+```
+
+## Using the Statsig Provider with the OpenFeature SDK
```
[the remark this answers is not in the document]
Comment (jenshenneberg) [reply src/OpenFeature.Contrib.Providers.Statsig/README.md:41]: It's a great idea, but as you mention it's probably more for the [dotnet-sdk](https://github.com/open-feature/dotnet-sdk) repo, or maybe for this repo in another PR.

File: test/OpenFeature.Contrib.Providers.Statsig.Test/EvaluationContextExtensionsTests.cs (hunk 13 above)
Comment (jenshenneberg) [reply test/OpenFeature.Contrib.Providers.Statsig.Test/EvaluationContextExtensionsTests.cs:96]: The Statsig API handles the null user edge case (by returning false for everything) and I haven't thought of any reason as to diverge from this. But I certainly can throw an exception instead if you think that is preferable.

File: src/OpenFeature.Contrib.Providers.Statsig/StatsigProviderException.cs (hunk 11 above)
Comment (jenshenneberg) [reply src/OpenFeature.Contrib.Providers.Statsig/StatsigProviderException.cs:8]: Turns out there is an [FeatureProviderException](https://github.com/open-feature/dotnet-sdk/blob/main/src/OpenFeature/Error/FeatureProviderException.cs) 😄

File: src/OpenFeature.Contrib.Providers.Statsig/README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (jenshenneberg) [reply src/OpenFeature.Contrib.Providers.Statsig/README.md]: Example update 😄

File: src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs (hunk 7 above)
Comment (jenshenneberg) [reply src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs:64]: Statsig does support these operations, and there is an example [from the corresponding java implementation](https://github.com/open-feature/java-sdk-contrib/blob/main/providers/statsig/src/main/java/dev/openfeature/contrib/providers/statsig/StatsigProvider.java) on how to do it. However, I've chosen not to for a couple of reasons:
- I've gone for a minimal implementation
- We don't need this in phocas (My preference is that a feature system should return true/false to avoid it being used as a config system)
- I don't know the Layer functionality in Statsig, so would have to familiarise myself with it.

File: src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs (hunk 5 above)
Comment (jenshenneberg) [reply src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs:58]: It does give me a `CS0029	Cannot implicitly convert type 'OpenFeature.Model.ResolutionDetails<bool>' to 'System.Threading.Tasks.Task<OpenFeature.Model.ResolutionDetails<bool>>'	OpenFeature.Contrib.Providers.Statsig (net462), OpenFeature.Contrib.Providers.Statsig (net5.0), OpenFeature.Contrib.Providers.Statsig (net6.0), OpenFeature.Contrib.Providers.Statsig (net7.0), OpenFeature.Contrib.Providers.Statsig (netstandard2.0)	C:\repos\dotnet-sdk-contrib\src\OpenFeature.Contrib.Providers.Statsig\StatsigProvider.cs	58	Active
`

Comment (jenshenneberg) [reply src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs:58]: Cleaner code, but now has a warning. It' a good point, but I think I'll stick with the original

File: src/OpenFeature.Contrib.Providers.Statsig/README.md (hunk 4 above)
Comment (jenshenneberg) [reply src/OpenFeature.Contrib.Providers.Statsig/README.md:90]: - I didn't want to add userId, as it'll be removed and replaced by targeting key as soon as [this](https://github.com/open-feature/dotnet-sdk/pull/231) is released. There is comment in the provider about it. 

- The other fields mentioned in the documentation that are in the above list, are the ones that are not (yet?) present in the [StatsigUser](https://github.com/statsig-io/dotnet-sdk/blob/main/dotnet-statsig/src/Statsig/StatsigUser.cs). So I have nowhere to map them.

- Good point on the extension - I've had that thought myself, but decided to leave it for the next iteration

File: src/OpenFeature.Contrib.Providers.Statsig/EvaluationContextExtensions.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (jenshenneberg) [reply src/OpenFeature.Contrib.Providers.Statsig/EvaluationContextExtensions.cs:16]: I've decided to leave it as is and revisit the logic when [this](https://github.com/open-feature/dotnet-sdk/pull/231) is released.

**Review by LAttenborough — COMMENTED** [review]
I don't have anything to add that hasnt already been mentioned by the others. Ill leave this for those that have requested changes to actually approve.

The minimal implementation for the POC that Stardust is doing makes sense. I like the idea of keeping this as a way to retrieve boolean values only 👍

File: src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs (hunk 9 above)
Comment (jenshenneberg) [reply src/OpenFeature.Contrib.Providers.Statsig/StatsigProvider.cs]: I ended op going for a simpler approach to the provider class by using the Statsig server driver instead. The idea is to leave the scoping up to where we've adding Featureflagging to the application (Should probably be added to dependency injection as a singleton). I still added a concurrency test of sorts.

File: src/OpenFeature.Contrib.Providers.Statsig/EvaluationContextExtensions.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jenshenneberg) [reply src/OpenFeature.Contrib.Providers.Statsig/EvaluationContextExtensions.cs]: I'd prefer to leave it as is in the tests and then fix it when I can fix [this](https://github.com/phocassoftware/openfeature-dotnet-sdk-contrib/blob/11f24c347645d04ccf3ef34f61567bbf9ea8ee0a/src/OpenFeature.Contrib.Providers.Statsig/EvaluationContextExtensions.cs#L26) todo
